File: frontend/src/lib/auth/pkce.ts
/**
 * PKCE(Proof Key for Code Exchange) 유틸.
 * 인가 코드가 프론트를 잠깐 거치므로 code_verifier/code_challenge 로 코드 탈취를 막는다.
 * (plan.md: 프론트가 인가를 시작, client_secret 은 백엔드에만)
 */

function base64UrlEncode(bytes: Uint8Array): string {
  let binary = "";
  for (const byte of bytes) {
    binary += String.fromCharCode(byte);
  }
  return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

function randomBytes(length: number): Uint8Array {
  const bytes = new Uint8Array(length);
  crypto.getRandomValues(bytes);
  return bytes;
}

/** code_verifier: 43~128자 base64url 무작위 문자열 (32바이트 → 43자) */
export function generateCodeVerifier(): string {
  return base64UrlEncode(randomBytes(32));
}

/** code_challenge: code_verifier 를 SHA-256 해시 후 base64url */
export async function generateCodeChallenge(verifier: string): Promise<string> {
  const data = new TextEncoder().encode(verifier);
  const digest = await crypto.subtle.digest("SHA-256", data);
  return base64UrlEncode(new Uint8Array(digest));
}

/** CSRF 방지용 state 값 */
export function generateState(): string {
  return base64UrlEncode(randomBytes(16));
}

File: mobile/lib/core/auth/pkce.dart
import 'dart:convert';
import 'dart:math';

import 'package:crypto/crypto.dart';

/// PKCE 값 생성(사양서 §2.2.1). 웹과 규격이 같다.
///
/// `code_challenge_method` 는 항상 `S256` 이다. 서버는 `code_verifier` 를 그대로 제공자에게
/// 넘길 뿐 형식을 재해석하지 않으므로, 여기서 만든 값이 곧 제공자에게 도착한다.
abstract final class Pkce {
  static final Random _random = Random.secure();

  /// 무작위 32바이트 → base64url(패딩 제거) = 43자.
  static String verifier() => _base64Url(_randomBytes(32));

  /// `SHA-256(verifier)` → base64url.
  static String challenge(String verifier) =>
      _base64Url(sha256.convert(utf8.encode(verifier)).bytes);

  /// 무작위 16바이트 → base64url = 22자.
  static String state() => _base64Url(_randomBytes(16));

  static List<int> _randomBytes(int length) =>
      List<int>.generate(length, (_) => _random.nextInt(256));

  static String _base64Url(List<int> bytes) =>
      base64Url.encode(bytes).replaceAll('=', '');
}
